include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+ * with the reader_writer_latch as either a reader or writer in a user friendly RAII wrapper.
+ *
+ * The reader_writer_latch class is move-only and cannot be default constructed.
+ */
+class reader_writer_latch
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    reader_writer_latch()                                        = delete;  //!< Deleted.
+    reader_writer_latch(reader_writer_latch const &)             = delete;  //!< Deleted.
+    reader_writer_latch(reader_writer_latch &&)                  = default; //!< Defaulted.
+    reader_writer_latch & operator=(reader_writer_latch const &) = delete;  //!< Deleted.
+    reader_writer_latch & operator=(reader_writer_latch &&)      = default; //!< Defaulted.
+    ~reader_writer_latch()                                       = default; //!< Defaulted.
+
+    /*!\brief Constructs the reader_writer_latch with the reader count, writer count and the associated data structure.
+     * \tparam concurrent_t The type of the concurrent data structure. Must have a member function `close`.
+     * \param rcount The expected number of consumer threads.
+     * \param wcount The expected number of producer threads.
+     * \param ds     The concurrent data structure.
+     *
+     * \details
+     *
+     * Initialises a std::contrib::latch for the consumer threads and a std::contrib::latch for the producer threads.
+     * Initialises the completion function which calls `ds.close()` on completion. Note, only a reference to the
+     * concurrent data structure is stored. The caller must ensure that `ds` is not destructed before this.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable concurrent data structure.
+    reader_writer_latch(reader_count const rcount, writer_count const wcount, concurrent_t & ds) :
+        reader_latch{static_cast<ptrdiff_t>(rcount.get())},
+        writer_latch{static_cast<ptrdiff_t>(wcount.get())},
+        completion_fn{[&ds] () { ds.close(); }}
+    {
+        assert(rcount.get() >= 0);
+        assert(wcount.get() >= 0);
+    }
+
+    /*!\brief Constructs the reader_writer_latch with the writer count and the associated data structure.
+     * \tparam concurrent_t The type of the concurrent data structure. Must have a member function `close`.
+     * \param wcount The expected number of producer threads.
+     * \param ds     The concurrent data structure.
+     *
+     * \details
+     *
+     * Delegates to reader_writer_latch::reader_writer_latch(reader_count, writer_count, concurrent_t) with reader
+     * count set to 0.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable concurrent data structure.
+    reader_writer_latch(writer_count const wcount, concurrent_t & ds) :
```

Comment:
That's true, it becomes a normal latch, except that a normal latch (by proposal) does not have a completion phase and I didn't want to also implement the barrier, because it is more complex. I am still hoping that this will get into c++20. But in general you are right, the interface is wonky and should only allow writers and readers as the class name suggests.